File: K次取反后最大化的数组和.cpp
class Solution {
public:
    int largestSumAfterKNegations(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int minusCnt = 0;
        int sum = 0;
        for(int i = 0; i < nums.size(); i++){
            if(nums[i] < 0)
                minusCnt++;
            sum += nums[i];
        }
        for(int i = 0; i < min(k, minusCnt); i++){ // 先把负的变成正的
            nums[i] = -nums[i];
            sum += 2 * nums[i];
        }
        if(k > minusCnt){ //负的变成正的还没完
            k -= minusCnt;
            sort(nums.begin(), nums.end());
            if(k % 2 == 0) //最终不变
                return sum;
            else //把最小的变成负的
                return sum - 2 * nums[0];
        }
        else
            return sum;

    }
};

File: K次取反后最大化的数组和.py
class Solution:
    def largestSumAfterKNegations(self, nums: List[int], k: int) -> int:
        numsCnt = Counter(nums);
        total = sum(nums);

        for i in range(-100, 0): # change minus first
            if numsCnt[i]:
                cnt = min(k, numsCnt[i]);
                total += -i * 2 * cnt;
                k -= cnt;
                numsCnt[i] -= cnt;
                numsCnt[-i] += cnt; # prepare to find min when all minus are turned
                if k == 0:
                    return total;
        if not numsCnt[0] and k % 2 != 0:
            for i in range(1, 101):
                if numsCnt[i]:
                    total -= i * 2;
                    return total;

        return total;
